src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+from __future__ import print_function
+
+import attr
+import os
+
+from hypothesis import (
+    given,
+)
+from hypothesis.strategies import (
+    binary,
+    lists,
+)
+
+from twisted.python.filepath import (
+    FilePath,
+)
+
+from twisted.internet import defer
+
+from testtools.matchers import (
+    Not,
+    Equals,
+    Always,
+    HasLength,
+    MatchesStructure,
+)
+from testtools.twistedsupport import (
+    succeeded,
+)
+from testtools import (
+    ExpectedException,
+)
+
+from eliot import (
+    Message,
+)
+
+from magic_folder.magic_folder import (
+    LocalSnapshotService,
+    LocalSnapshotCreator,
+)
+from magic_folder.snapshot import (
+    create_local_author,
+)
+from magic_folder.magicpath import (
+    path2magic,
+)
+from .. import magicfolderdb
+
+from .common import (
+    SyncTestCase,
+)
+from .strategies import (
+    path_segments,
+)
+
+@attr.s
+class MemorySnapshotCreator(object):
+    """
+    A way to test LocalSnapshotService with an in-memory database.
+    """
+    processed = attr.ib(default=attr.Factory(dict))
+    def process_item(self, path):
+        Message.log(
+            message_type=u"memory-snapshot-creator:process_item",
+            path=path.asTextMode("utf-8").path,
+        )
+        if path not in self.processed.keys():
+            # record number of parents
+            self.processed[path] = 0
+        else:
+            num_parents = self.processed[path]
+            self.processed[path] = num_parents + 1
+        return defer.succeed(None)
+
+
+class LocalSnapshotServiceTests(SyncTestCase):
```

Comment:
A big advantage of splitting the service from the snapshot creator is that it makes testing easier.  A lot of the tests in this class that have to do with creating snapshots should move to the `LocalSnapshotCreator` test case (or else they are already redundant with tests there).  The tests for `LocalSnapshotService`  Can be much simpler.  The service doesn't know anything about snapshots or parents and has zero logic to dealing with either.  It can be completely tested without a database, without an author, etc.

`MemorySnapshotCreator` can really literally just append to a list to prove the service is behaving correctly.  The list should always be equal to exactly the valid paths that have been passed to `add_file`, in the order they were passed.